File: examples/python/common/bootstrap.py
"""Bootstrap helper for Python examples in the Bitcoin Sandbox."""
from typing import Tuple
from .rpc import BitcoinRPC, get_config


def bootstrap_lab(rpc: BitcoinRPC = None) -> Tuple[BitcoinRPC, BitcoinRPC]:
    """
    Ensures the 'lab' wallet exists, is loaded, and has spendable funds
    (by mining 101 blocks to clear coinbase maturity if necessary).
    
    Returns:
        (node_rpc, wallet_rpc): Pair of RPC clients for node-level and wallet-level calls.
    """
    config = get_config()
    wallet_name = config["rpc_wallet"]
    
    if rpc is None:
        rpc = BitcoinRPC()

    # 1. Verify/Load or Create the lab wallet
    loaded_wallets = rpc.call("listwallets")
    if wallet_name not in loaded_wallets:
        # Check if wallet directory exists on disk
        wallet_dir_info = rpc.call("listwalletdir")
        existing_wallets = [w["name"] for w in wallet_dir_info.get("wallets", [])]
        if wallet_name in existing_wallets:
            rpc.call("loadwallet", [wallet_name])
        else:
            # Create descriptor wallet
            rpc.call("createwallet", [wallet_name])

    wallet_rpc = rpc.for_wallet(wallet_name)

    # 2. Check spendable balance; mine 101 blocks if 0 (coinbase maturity)
    balance = wallet_rpc.call("getbalance")
    if balance == 0:
        mining_addr = wallet_rpc.call("getnewaddress", ["bootstrap_mining", "bech32m"])
        wallet_rpc.call("generatetoaddress", [101, mining_addr])

    return rpc, wallet_rpc

File: examples/typescript/src/common/bootstrap.ts
import { BitcoinRPC, getConfig } from "./rpc.js";

export interface BootstrapResult {
  rpc: BitcoinRPC;
  walletRpc: BitcoinRPC;
}

export async function bootstrapLab(rpc?: BitcoinRPC): Promise<BootstrapResult> {
  const config = getConfig();
  const walletName = config.rpcWallet;
  const nodeRpc = rpc || new BitcoinRPC();

  // 1. Verify/Load or Create the lab wallet
  const loadedWallets = await nodeRpc.call<string[]>("listwallets");
  if (!loadedWallets.includes(walletName)) {
    const walletDirInfo = await nodeRpc.call<{ wallets: { name: string }[] }>("listwalletdir");
    const existingWallets = (walletDirInfo.wallets || []).map((w) => w.name);
    if (existingWallets.includes(walletName)) {
      await nodeRpc.call("loadwallet", [walletName]);
    } else {
      await nodeRpc.call("createwallet", [walletName]);
    }
  }

  const walletRpc = nodeRpc.forWallet(walletName);

  // 2. Check spendable balance; mine 101 blocks if 0 (coinbase maturity)
  const balance = await walletRpc.call<number>("getbalance");
  if (balance === 0) {
    const miningAddr = await walletRpc.call<string>("getnewaddress", ["bootstrap_mining", "bech32m"]);
    await walletRpc.call("generatetoaddress", [101, miningAddr]);
  }

  return { rpc: nodeRpc, walletRpc };
}
